src/main/java/com/uber/profiling/reporters/GraphiteOutputReporter.java

```diff
@@ -0,0 +1,172 @@
+        .replaceAll("\\.", "-");
+    String newPrefix = String.join(".", prefix, tag, appId, host, process);
+
+    Map<String, Object> formattedMetrics = getFormattedMetrics(metrics);
+    long timestamp = System.currentTimeMillis() / 1000;
+    for (Map.Entry<String, Object> entry : formattedMetrics.entrySet()) {
+      out.printf(
+          newPrefix + "." + entry.getKey() + " " + entry.getValue() + " " + timestamp + "%n");
+    }
+  }
+
+  // Format metrics in key=value (line protocol)
+  private Map<String, Object> getFormattedMetrics(Map<String, Object> metrics) {
+    Map<String, Object> formattedMetrics = new HashMap<>();
+    for (Map.Entry<String, Object> entry : metrics.entrySet()) {
+      String key = entry.getKey();
+      Object value = entry.getValue();
+      logger.debug("Raw Metric-Name = " + key + ", Metric-Value = " + value);
+      if (value != null && value instanceof List) {
+        List listValue = (List) value;
+        if (!listValue.isEmpty() && listValue.get(0) instanceof String) {
+          List<String> metricList = (List<String>) listValue;
+          formattedMetrics.put(key, String.join(",", metricList));
+        } else if (!listValue.isEmpty() && listValue.get(0) instanceof Map) {
+          List<Map<String, Object>> metricList = (List<Map<String, Object>>) listValue;
+          int num = 1;
+          for (Map<String, Object> metricMap : metricList) {
+            String name = null;
+            if (metricMap.containsKey("name") && metricMap.get("name") != null && metricMap
+                .get("name") instanceof String) {
+              name = (String) metricMap.get("name");
+              name = name.replaceAll("\\s", "");
+            }
+            for (Map.Entry<String, Object> entry1 : metricMap.entrySet()) {
+              if (StringUtils.isNotEmpty(name)) {
+                formattedMetrics.put(key + "." + name + "." + entry1.getKey(), entry1.getValue());
+              } else {
+                formattedMetrics.put(key + "." + entry1.getKey() + "." + num, entry1.getValue());
+              }
+            }
+            num++;
+          }
+        }
+      } else if (value != null && value instanceof Map) {
+        Map<String, Object> metricMap = (Map<String, Object>) value;
+        for (Map.Entry<String, Object> entry1 : metricMap.entrySet()) {
+          String key1 = entry1.getKey();
+          Object value1 = entry1.getValue();
+          if (value1 != null && value1 instanceof Map) {
+            Map<String, Object> value2 = (Map<String, Object>) value1;
+            int num = 1;
+            for (Map.Entry<String, Object> entry2 : value2.entrySet()) {
+              formattedMetrics
+                  .put(key + "." + key1 + "." + entry2.getKey() + "." + num, entry2.getValue());
+            }
+            num++;
+          }
+        }
+      } else {
+        formattedMetrics.put(key, value);
```

Comment:
This code path will be hit when value == null? Suggest checking value == null before checking it type.

Reply:
Hi，@buy-uber , Thanks for your review.

I just refactored the code, and add a test case.
Can you review my code again?

Many Thanks.

Reply:
Cool, thanks for contributing here and making the refactoring! It looks good to me now. Let me merge your PR.